### PR TITLE
Feature: Bookmark Categories

### LINKED ISSUE #5922: Bookmarking categories
### What is the user problem or growth opportunity you want to see solved?

It seems like with the release that's just out and will probably soon be available in the PlayStore, the category search [will work again](https://github.com/commons-app/apps-android-commons/pull/5830) in the app. Once category support is back it would be really great if one could also [bookmark](https://commons.wikimedia.org/wiki/File:20240529_xl_09-Commons_Android_App_screenshots_user_interface_5.0.1~da0b2c28e_May_2024.png) categories, not just files.

I think it would be far more useful than bookmarking files. For example, I'd like to bookmark [this category](https://commons.wikimedia.org/wiki/Category:Spoken_Wikipedia_articles_using_English-language_speech_synthesis) so I can use the app to quickly and conveniently and with dark mode browse through spoken Wikipedia audios to download into my podcast player app (support for playing audio files is a separate issue at #5915)

### How do you know that this problem exists today? Why is this important?

It would be very useful to bookmark categories. One could bookmark all sorts of categories one is interested in. Maybe one could even include some of the bookmarked categories into some Home explore feed, e.g. showing the most recent or a few random items for a few categories bookmarked and added to the Feed. If needed I could list more examples of how people find bookmarking categories useful.

### Who will benefit from it?

People who use the app to explore, discover, search and consume media.

### Anything else you would like to add?

In the bookmarks page at some point there could be two tabs: one for bookmarked files and one for bookmarked categories. It's not necessary to include in the initial version, e.g. I don't bookmark any files but would bookmark several categories. If one has bookmarked many files accessing the bookmarked categories quickly which is one key thing the bookmarks would be good for would diminish as one has to first scroll and search long which is why having a separate tab would be good for people who bookmark both types of pages.

### PR DESCRIPTION
**Pull request by parneet-guraya**
Fixes #5922

What changes did you make and why?

- Add bookmark button in bookmark details screen.
- Add new bookmark tab in bookmarks screen.

**Tests performed (required)**

Tested betaDebug on Oneplus 9RT 5G (Android 14)

**Screenshots (for UI changes only)**

https://github.com/user-attachments/assets/221c73ef-fb16-4225-9064-515a3f3aedf2

### CODE REVIEW

**Comment by parneet-guraya** [comment]
@nicolas-raoul 

Hi, I have a question, right now we're loading the entire data. But, do we want to load the items lazily (paging) ? Other bookmarks types (picture, items, locations) does not use paging either. 

Obviously that depends on the size of data but do we expect the bookmarks to be in large amounts?

Thanks :)

**Comment by nicolas-raoul** [comment]
Show when logged out: yes

Paging, sorting, searching: Great idea, though the added complexity would currently not be justified. Would you mind creating a new issue and labeling it as enhancement, low priority? Thanks a lot! 🙂

**Comment by parneet-guraya** [comment]
> Show when logged out: yes

Great! I also kept it that way.

> 
> Paging, sorting, searching: Great idea, though the added complexity would currently not be justified. Would you mind creating a new issue and labeling it as enhancement, low priority? Thanks a lot! 🙂

I will shortly, outside right now (commenting from my mobile). :)

File: app/src/main/java/fr/free/nrw/commons/category/CategoryDetailsViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import androidx.lifecycle.viewModelScope
+import fr.free.nrw.commons.bookmarks.category.BookmarkCategoriesDao
+import fr.free.nrw.commons.bookmarks.category.BookmarksCategoryModal
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+class CategoryDetailsViewModel(
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/category/CategoryDetailsViewModel.kt:14]: Can you add kdoc to all new classes? And ideally functions too. Thanks! :-)

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

File: app/src/main/java/fr/free/nrw/commons/category/CategoryDetailsViewModel.kt (hunk 1 above)
Comment (parneet-guraya) [reply app/src/main/java/fr/free/nrw/commons/category/CategoryDetailsViewModel.kt:14]: My bad :(  I was going to but forgot. Doing it now.

**Comment by parneet-guraya** [comment]
Docs added :)

**Review by nicolas-raoul — APPROVED** [review]
Working great, thanks a lot!

**Comment by parneet-guraya** [comment]
@nicolas-raoul This closes this #5318 issue as well. Take a look :)

**Pull request merged**